Infrastructure setup

ekiefl opened the pull request:
This PR strips poetry from the project and replaces it with uv.

Additionally:

* Python version bounds include 3.10-3.13. 3.9 is EOL, and 3.14 may have some compatibility issues with some of our libraries. So to be conservative, I've set the upper bound to 3.13, though we will experiment with 3.14 once our package has content.
* Enables CI tests
	- lint has been split into linting and typechecking. This is because linting doesn't require package dependencies, but typechecking, which requires full abstract syntax trees of all dependencies, does. By splitting these actions up, we speed up linting by decoupling it from a complex installation.
	- I created a test matrix spanning linux/windows/mac that's tested over our lower/upper bounds, totaling 6 os+python combinations.
	- For all CI actions, we cache the installation state, reducing runtimes from 5m to seconds. The cache hits if uv.lock hasn't changed, and misses and re-caches if it has.
* Our documentation pages now properly render. I populated the installation page, which has tabulated installation instructions depending on OS and package manager (pip/uv).

---

Once this merges we're ready to start porting code into the repo.

ekiefl commented:
I also just noticed you didn't update the title in `_variables.yml`

<img width="1506" height="272" alt="image" src="https://github.com/user-attachments/assets/ed717ee9-99fa-40ba-96ae-7487c018c6e5" />

ekiefl committed:
Changes requested in https://github.com/Arcadia-Science/glass-box-umap/pull/1

ekiefl committed:
Move deps out of conda yaml and into pyproject.toml

- Excluding torch deps for now

ekiefl committed:
Turn on tests

ekiefl committed:
Type check on pre-commit

ekiefl commented:
CI environment properly runs (with poetry)

ekiefl committed:
Move from flat to src layout

ekiefl committed:
Remove conda envs dir

ekiefl committed:
Switch from poetry to uv

ekiefl committed:
Add live doc watching

ekiefl committed:
Update README to reflect poetry -> uv

ekiefl committed:
Track python version

- This will be useful for consistent dev/docs environments between
  developers
- We'll protect against forward compatibility bias with our CI test
  matrix for earlier python versions

ekiefl committed:
Restrict 3.14 for now

ekiefl committed:
Trial some new CI actions

ekiefl committed:
No apparent need for these

ekiefl committed:
Upgrade and then pin dev dependencies

ekiefl committed:
Split lint and type check

- Type check needs the *whole* dep collection but linting only needs the
  dev deps
- So we use the power of uv dependency groups + github action caching to
  optimize as best we can

ekiefl committed:
Use single source of truth for python version

ekiefl committed:
Avoid missing deps being added at runtime

ekiefl committed:
See if this fixes a 400 error due to race condition?

ekiefl committed:
Trial key rotation for uv env caching

ekiefl committed:
Bump uv-setup version and add cache size debug step

ekiefl committed:
Look for cache in right place

ekiefl committed:
Move cache size check to after dependencies are installed

ekiefl committed:
Remove temporary cache size check

ekiefl committed:
Remove version. Have it be latest

ekiefl committed:
Add comment and option to add more glob patterns

ekiefl committed:
Prune dependencies

ekiefl committed:
Update test matrix yml

ekiefl committed:
Update lint.yml

ekiefl committed:
Add --no-sync to prevent re-installing

Was getting in test.yml:

Removed virtual environment at: .venv
Creating virtual environment at: .venv

Hopefully this prevents that.

ekiefl commented:
CI environment properly runs (with uv)

* Environments are task specific (e.g. lint builds environment with only the dev deps)
* Environment caches properly hit and properly miss upon uv.lock changes.

ekiefl committed:
Add torch as a dep

- Lay out the installation language in install.md

ekiefl committed:
Touch up readme

ekiefl committed:
Rename

james-golden-arcadia commented:
After I run the curl command to install uv, I get this when I try to run uv sync:
```
(base) ubuntu@ip-172-31-41-173:~/james/package$ uv sync --group dev
Command 'uv' not found, but can be installed with:
sudo snap install astral-uv
```

Running the snap command works, but so does:
```
(base) ubuntu@ip-172-31-41-173:~/james/package$ export PATH="$HOME/.local/bin:$PATH"
```

I think everyone will know to do this, but wanted to mention in case we should add it to README.

james-golden-arcadia commented:
The unit test passes!

I just had to acitvate the env or run
```
uv run make docs
```

and then the docs were created as expected. Maybe we should make this more explicit in README?

ekiefl committed:
Clarify language in README.md

james-golden-arcadia reviewed (approved):
Thanks for the README update - LGTM

ekiefl commented:
> I think everyone will know to do this, but wanted to mention in case we should add it to README

Added a remark about verifying that uv is installed and available. I kept it generic and avoided any specific error messages and their solutions. How someone installs their package manager is outside the responsibility of any given package, so best not to be opinionated.

> and then the docs were created as expected. Maybe we should make this more explicit in README?

Good catch. Replaced `make docs` with `uv run make docs`

Changes in [b55f239](https://github.com/Arcadia-Science/glass-box-umap/pull/2/commits/b55f2399400e1cd7c5b8611fa2306b6538633a95)

ekiefl commented:
Ok great. There'll be more infrastructure to lay down once we get some of the functionality in. But for now, I think I'll merge this. I think it's ready for you to port some of the code over!
